### ISSUE TITLE
What's the correct way to use allgather?

### ISSUE THREAD

**Issue by yuzhms** [horovod/horovod#2027]
**Environment:**
1. Framework: Pytorch
2. Framework version:
3. Horovod version:
4. MPI version:
5. CUDA version:
6. NCCL version:
7. Python version:
8. OS and version:
9. GCC version:

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Your question:**
I'am implementing a loss which need to collect data across ranks. I found directly using horovod is not capable to do this because data is not shared across rank. I also found `allgather` may be the thing I needed and I try to do something like this:

```python

# data
dataset = my_data()
sampler = DistributedSampler(dataset , num_replicas=hvd.size(), rank=hvd.rank())
loader = DataLoader(dataset , batch_size=batch_size, sampler=sampler, pin_memory=True,
                                 drop_last=True, num_workers=workers)
# model
model = my_model()
model.cuda()
optimizer = torch.optim.Adam(model.parameters(), lr)
hvd.broadcast_parameters(model.state_dict(), root_rank=0)
hvd.broadcast_optimizer_state(optimizer, root_rank=0)
optimizer = hvd.DistributedOptimizer(optimizer, named_parameters=model.named_parameters())

# loss
loss_func = my_loss()

# train
for data in loader:
  x, label = data
  x = x.cuda()
  label = label.cuda()
  y= model(x)
  y = hvd,.llgather(y)
  loss = loss_func(y)
  loss.backward()
  optimizer.synchronize()
  
```

I expect such implementation is the same as I use `torch.nn.DataParallel` as below:

```python
# data
dataset = my_data()
loader = DataLoader(dataset , batch_size=batch_size, pin_memory=True,
                                 drop_last=True, num_workers=workers)
# model
model = my_model()
model = torch.nn.DataParallel(model)
model.cuda()
optimizer = torch.optim.Adam(model.parameters(), lr)

# loss
loss_func = my_loss()

# train
for data in loader:
  x, label = data
  x = x.cuda()
  label = label.cuda()
  y = model(x)
  loss = loss_func(y)
  loss.backward()
```

I would like to know do I use ` allgather`  correctly? If not, how to modify it?

**Comment by tgaddair** [comment]
Hey @yuzhms, can you explain the purpose of the allgather here?  In most cases, using Horovod's `DistributedOptimizer` will give you the same result as using `torch.nn.DataParallel` without needing to do any custom operations.  

You certainly can use `hvd.allgather` if you need to, but be aware that it will result in a tensor containing all the tensors across all workers stacked together.

**Comment by yuzhms** [comment]
The purpose is I am implementing a loss which need to collect data across ranks. @tgaddair 

For example, I have a batch of data which have the batch size of 1024, and I have 8 cards. In this case, I think I should set the batch size to 128 for each card. So if I calculate the loss, I will use 128 batch not the total 1024 batch.

In `torch.nn.DataParallel`, I can wrap my model by `DataParallel`, which use 128 batch for each card, and calculate the loss outside `DataParallel`, so I use the 1024 batch.

I would like to see the same thing can be done by Horovod, and I found `hvd.allgather` may be useful. But there is rarely documentation talking about how to use it.
